fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,779 @@
+    * (num_equations x dim)
+    * @param[in] maxE maximum characteristic speed (eigenvalue of flux jacobian)
+    * @param[in] nor normal vector (not a unit vector) (dim x 1)
+    * @param[out] flux ½(F(u⁺,x)n + F(u⁻,x)n) - ½λ(u⁺ - u⁻)
+    */
+   void Eval(const Vector &state1, const Vector &state2, const Vector &fluxN1,
+             const Vector &fluxN2, const double maxE, const Vector &nor,
+             Vector &flux)
+   {
+      // NOTE: nor in general is not a unit normal
+
+      flux = state1;
+      flux -= state2;
+      // here, sqrt(nor*nor) is multiplied to match the scale with fluxN
+      flux *= maxE * sqrt(nor * nor);
+      flux += fluxN1;
+      flux += fluxN2;
+      flux *= 0.5;
+   }
+};
+
+// Upwind Flux, Not Yet Implemented
+class UpwindFlux : public RiemannSolver
+{
+public:
+   // Upwind Flux, Not Yet Implemented
+   void Eval(const Vector &state1, const Vector &state2, const Vector &fluxN1,
+             const Vector &fluxN2, const double maxE, const Vector &nor,
+             Vector &flux)
+   {
+      // NOTE: nor in general is not a unit normal
+      mfem_error("Not Implemented");
+   }
+};
+
+class AdvectionFormIntegrator : public HyperbolicFormIntegrator
+{
+private:
+   VectorCoefficient &b;  // velocity coefficient
+   Vector bval;           // velocity value storage
+
+public:
+   /**
+    * @brief Compute F(u)
+    *
+    * @param U U (u) at current integration point
+    * @param Tr current element transformation with integration point
+    * @param FU F(u) = ubᵀ
+    * @return double maximum characteristic speed, |b|
+    */
+   double ComputeFlux(const Vector &U, ElementTransformation &Tr,
+                      DenseMatrix &FU)
+   {
+      b.Eval(bval, Tr, Tr.GetIntPoint());
+      MultVWt(U, bval, FU);
+      return bval.Norml2();
+   }
+   /**
+    * @brief Compute normal flux, F(u)n
+    *
+    * @param U U (u) at current integration point
+    * @param normal normal vector, usually not a unit vector
+    * @param Tr current element transformation with integration point
+    * @param FUdotN F(u)n = u(b⋅n)
+    * @return double maximum characteristic speed, |b|
+    */
+   double ComputeFluxDotN(const Vector &U, const Vector &normal,
+                          ElementTransformation &Tr, Vector &FUdotN)
+   {
+      b.Eval(bval, Tr, Tr.GetIntPoint());
+      const double bN = bval * normal;
+      FUdotN = U;
+      FUdotN *= bN;
+      return bval.Norml2();
+   }
+
+   /**
+    * @brief Construct a new Advection Element Form Integrator object with given
+    * integral order offset
+    *
+    * @param[in] rsolver_ numerical flux
+    * @param dim spatial dimension
+    * @param b_ velocity coefficient, possibly depends on space
+    * @param IntOrderOffset_ 2*p + IntOrderOffset will be used for quadrature
+    */
+   AdvectionFormIntegrator(RiemannSolver *rsolver_, const int dim,
+                           VectorCoefficient &b_,
```

Comment:
I think the dimension can be inferred from the coefficient using `VectorCoefficient::GetVDim`